move to explicit avx2 switching

EscapeZero opened the pull request:
Summary:
The main idea here is to try and give build system more flexibility on when various AVX instruction are defined, previously it was based solely on compiler defined preprocessor flags.

Here we re-use `CPU_CAPABILITY` which already needs to be defined for each pass in `["DEFAULT", "AVX", "AVX2"]` over the source files.

To give a slightly more concrete reason this is needed, is that we have not found a way to override `/arch` flags previously specified on the command line from visual studio (causing us to duplicate symbols in some cases).

Test Plan: CI green

Differential Revision: D21218512

EscapeZero commented:
@malfet my PR has revealed that `pytorch_bazel_test` compiles `DEFAULT, AVX, AVX2` but does not actually enable the various instruction sets for each of the passes. I would expect `-mavx, -mavx2` to be set here: 
https://github.com/pytorch/pytorch/blob/c306f2ed0832e0545228e779d3dce2ded2153566/BUILD.bazel#L623-L632

ezyang commented:
@EscapeZero I feel that there may be some confusion about how these files are supposed to work. The vec256 headers are meant to be compiled multiple times with different settings of AVX; this is how we produce multiple copies of code built under different vectorization settings, so we can support a variety of CPUs when we actually run them at runtime. As such, it's appropriate for vec256 to match on the compiler provided AVX defs, because that will indeed tell you what instruction set you should be compiling with.

You can see that the Bazel build system builds the native/cpu/ files multiple times in https://github.com/pytorch/pytorch/blob/c306f2ed0832e0545228e779d3dce2ded2153566/aten.bzl Although... I don't see the avx flags, so maybe these rules are not written correctly. cc @malfet

malfet commented:
@EscapeZero @ezyang yes, bazel builds needs to be fixed.

EscapeZero commented:
@ezyang I get that it is supposed to be compiled multiple times under each arch. The more important thing that this diff enables is described at the end of my summary (a mode file passes /arch:AVX2 to all compilation units) which makes it impossible for me to force vec256 into the #else section (since it relies on __AVX2__ instead of something we have direct control over).

Some random use cases:
What if you wanted to test the autovectorizer, by enabling avx2 but not directly calling intrinsics. That could be done after this diff.

dzhulgakov commented:
@ezyang - yeah, using CPU_CAPABILITY should be fine, we already rely on it in DispatchStub which is working with the same logic: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/DispatchStub.h#L195

So to me it looks ok.

malfet commented:
Fixed bazel build logic here: https://github.com/pytorch/pytorch/pull/37381

malfet commented on aten/src/ATen/cpu/vec256/vec256_bfloat16.h:
```diff
@@ -2,7 +2,7 @@
 #include <ATen/cpu/vec256/intrinsics.h>
 #include <ATen/cpu/vec256/vec256_base.h>
-#if defined(__AVX__) && !defined(_MSC_VER)
+#if defined(CPU_CAPABILITY_AVX2) && !defined(_MSC_VER)
```
This should be AVX | AVX2, isn't it?
```suggestion
#if (defined(CPU_CAPABILITY_AVX) || defined(CPU_CAPABILITY_AVX2)) && !defined(_MSC_VER)
```

EscapeZero replied:
So this was also a little fix I was sneaking in. If you look further down, the whole file contents are `ifdef` on `AVX2`, so the extra include is also only needed when `AVX2` is defined. But maybe this was intentional? maybe other headers rely on this to get sleef even though this file would not use it at all in the `AVX` case.

malfet commented on aten/src/ATen/cpu/vec256/vec256_qint.h:
```diff
@@ -101,7 +101,7 @@ inline void __attribute__((always_inline)) QuantizeAvx2(
     float inverse_scale,
     int64_t zero_point) {
-#if defined(__AVX2__) && defined(__FMA__)
+#if defined(CPU_CAPABILITY_AVX2) && defined(__FMA__)
```
Please remove `defined(__FMA__)` since `CAPABILITY_AVX2` implies that `fma` instruction is available
```suggestion
#if defined(CPU_CAPABILITY_AVX2)
```

malfet commented on aten/src/ATen/cpu/vec256/vec256_qint.h:
```diff
@@ -264,7 +264,7 @@ struct Vec256<c10::qint32> : public Vec256qi {
         Vec256<float> scale_zp_premul) const {
       __m256 float_vals = _mm256_cvtepi32_ps(vals);
-#if defined(__AVX2__) && defined(__FMA__)
+#if defined(CPU_CAPABILITY_AVX2) && defined(__FMA__)
```
Sam as above
```suggestion
#if defined(CPU_CAPABILITY_AVX2)
```

EscapeZero committed:
move to explicit avx2 switching (#37207)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37207

The main idea here is to try and give build system more flexibility on when various AVX instruction are defined, previously it was based solely on compiler defined preprocessor flags.

Here we re-use `CPU_CAPABILITY` which already needs to be defined for each pass in `["DEFAULT", "AVX", "AVX2"]` over the source files.

To give a slightly more concrete reason this is needed, is that we have not found a way to override `/arch` flags previously specified on the command line from visual studio (causing us to duplicate symbols in some cases).

Test Plan: CI green

Differential Revision: D21218512

fbshipit-source-id: ba2ca2a54679c68eda83e9bba8999f2d29e932c7
